packages/solarwinds-apm/src/init.ts

```diff
@@ -94,52 +83,51 @@ export function init() {
       return
     }
 
+    // initialize instrumentations before any asynchronous code
+    const registerInstrumentations = initInstrumentations(config)
+
     const resource = Resource.default()
-      .merge(
-        detectResourcesSync({
-          detectors: [hostDetectorSync, osDetectorSync, processDetectorSync],
-        }),
-      )
+      .merge(getDetectedResource())
       .merge(
         new Resource({
           [SemanticResourceAttributes.SERVICE_NAME]: config.serviceName,
         }),
       )
-
     const reporter = sdk.createReporter(config)
-    const tracerProvider = initTracing(
-      config,
-      resource,
+
+    oboe.debug_log_add((module, level, sourceName, sourceLine, message) => {
+      const logger = diag.createComponentLogger({
+        namespace: `sw/oboe/${module}`,
+      })
+      const log = oboeLevelToOtelLogger(level, logger)
+
+      if (sourceName && level > oboe.DEBUG_INFO) {
+        const source = { source: sourceName, line: sourceLine }
+        log(message, source)
+      } else {
+        log(message)
+      }
+    }, config.oboeLogLevel)
+
+    if (resource.asyncAttributesPending) {
+      await resource.waitForAsyncAttributes?.()
+    }
+    sdk.sendStatus(
       reporter,
-      packageJson.version,
+      await sdk.initMessage(resource, packageJson.version),
     )
-    const meterProvider = initMetrics(config, resource, reporter, logger)
-    initInstrumentations(config, tracerProvider, meterProvider)
+
+    const tracerProvider = await initTracing(config, resource, reporter)
+    const meterProvider = await initMetrics(config, resource, reporter, logger)
+    registerInstrumentations(tracerProvider, meterProvider)
   }
 }
 
-function initTracing(
+async function initTracing(
   config: ExtendedSwConfiguration,
   resource: Resource,
   reporter: oboe.Reporter,
-  version: string,
 ) {
-  oboe.debug_log_add((module, level, sourceName, sourceLine, message) => {
-    const logger = diag.createComponentLogger({
-      namespace: `sw/oboe/${module}`,
-    })
-    const log = oboeLevelToOtelLogger(level, logger)
-
-    if (sourceName && level > oboe.DEBUG_INFO) {
-      const source = { source: sourceName, line: sourceLine }
-      log(message, source)
-    } else {
-      log(message)
-    }
-  }, config.oboeLogLevel)
-
-  sdk.sendStatus(reporter, sdk.initMessage(resource, version))
```

Comment:
This isn't tracing specific, there's no reason not to have it in the parent scope and earlier than this